### PR TITLE
Preserve correctness of `import.meta.url` in bundled modules.

### PR DESCRIPTION
**Pull request by usergenic**
Note that in the case where an imported module's URL is identical to the bundle URL, no modification to the module code is needed before bundling it, but for all other cases where `import.meta` occurs somewhere in the document contents, we construct a module-scoped stand-in variable and swap other references to import.meta to use the stand-in, which is declared as an object-literal like `{...import.meta, url: new URL(relativeUrlToModule, import.meta.url)}`.

### CODE REVIEW

**Comment by usergenic** [comment]
Fixes https://github.com/Polymer/tools/issues/294

File: packages/bundler/src/es6-rewriter.ts [hunk 1]
```diff
@@ -106,20 +120,37 @@ export class Es6Rewriter {
...
+                  !document.parsedDocument.contents.includes('import.meta')) {
+                return document.parsedDocument.contents;
+              }
+
+              // We need to rewrite instances of `import.meta` in the document
+              // to preserve its location because `import.meta` is used and the
+              // URL has changed as a result of bundling.
+              const relativeUrl =
+                  ensureLeadingDot(this.bundler.analyzer.urlResolver.relative(
+                      this.bundle.url, id as ResolvedUrl));
```
Comment (rictic) [inline packages/bundler/src/es6-rewriter.ts]: Is the cast of `id` needed? It looks like it's a ResolvedUrl already

File: packages/bundler/src/es6-rewriter.ts [hunk 2]
```diff
@@ -106,20 +120,37 @@ export class Es6Rewriter {
...
+              const newAst = this._rewriteImportMetaToBundleMeta(
+                  document.parsedDocument.ast as babel.File, relativeUrl);
```
Comment (rictic) [inline packages/bundler/src/es6-rewriter.ts]: We could put this constraint into the type system if we added a type checking function that checks that `document` is a `Document<ParsedJavascriptDocument>`

File: packages/bundler/src/es6-rewriter.ts [hunk 3]
```diff
@@ -432,4 +463,52 @@ export class Es6Rewriter {
...
+        babel.variableDeclaration(  // leave this in for clang-format
+            'const',
+            [babel.variableDeclarator(
+                babel.identifier(bundledImportMetaName),
+                babel.objectExpression([
+                  babel.spreadProperty(babel.memberExpression(
+                      babel.identifier('import'), babel.identifier('meta'))),
+                  babel.objectProperty(
+                      babel.identifier('url'),
+                      babel.newExpression(
```
Comment (rictic) [inline packages/bundler/src/es6-rewriter.ts]: We want to get the `.href` property of this new url, as `import.meta.url` is a string

File: packages/bundler/src/test/import-meta-url_test.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      import {aUrl, myUrl as bUrl} from './subfolder/b.js';
+      const myMeta = import.meta;
+      export {aUrl, bUrl, myMeta};
+    `,
+  });
+
+  const aUrl = analyzer.resolveUrl('a.js')!;
+  const bUrl = analyzer.resolveUrl('subfolder/b.js')!;
+  const cUrl = analyzer.resolveUrl('c.js')!;
+  const dUrl = analyzer.resolveUrl('d.js')!;
```
[the remark this answers is not in the document]
Comment (usergenic) [reply packages/bundler/src/test/import-meta-url_test.ts]: No, thanks!  I was adding another test case and then gave up; forgot that was there.  Removed.

File: packages/bundler/src/es6-rewriter.ts [hunk 5]
```diff
@@ -432,4 +463,52 @@ export class Es6Rewriter {
         importSpecifier,
         {type: 'ImportSpecifier', imported: babel.identifier(exportName)});
   }
+
+  private _rewriteImportMetaToBundleMeta(
+      moduleFile: babel.File,
+      relativeUrl: FileRelativeUrl): babel.File {
+    // Generate a stand-in for any local references to import.meta...
+    // const __bundledImportMeta = {...__importMeta, url: __bundledImportURL};
```
[the remark this answers is not in the document]
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: I previously had a local variable stand-in just for import.meta that is now gone. Old comment. Fixed to import.meta! Thanks!

File: packages/bundler/src/es6-rewriter.ts (hunk 1 above)
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: You're right. Fixed.

File: packages/bundler/src/es6-rewriter.ts (hunk 2 above)
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: You're right and in fact this code probably should have more guarantees/verification of the type of document it is loading anyways. I'll add a TODO here for this.

File: packages/bundler/src/es6-rewriter.ts [hunk 6]
```diff
@@ -432,4 +463,52 @@ export class Es6Rewriter {
...
+    const bundledImportMetaName = '__bundledImportMeta';
+    const bundledImportMetaDeclaration =
```
[the remark this answers is not in the document]
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: I love this idea.  I even made a personal TODO note earlier today "Look up to see if there's a template string builder that would make this code cleaner."

I would like to consider bringing that approach to all of the ast production in bundler, but I'm going to hold off on a separate PR to bring that machinery in here.  Adding TODO.  It's a great idea and would make this code a lot better.

I should note that as of right now though, Bundler does not have any babel parsing in it and relies purely on Analyzer. Use of the template literal stuff would essentially require ensuring I either find a nice way to clone the Analyzer's babel parsing stack or rebuild it locally.

[the remark this answers is not in the document]
Comment (aomarks) [reply packages/bundler/src/es6-rewriter.ts]: SGTM

File: packages/bundler/src/es6-rewriter.ts (hunk 3 above)
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: DERP!  Really good catch.  Fixed.

**Comment by usergenic** [comment]
@rictic & @aomarks comments addressed

**Comment by usergenic** [comment]
Awww nuts.  Something probably in the package-lock has updated causing something to break during gulp-typescript... https://ci.appveyor.com/project/Polymer/tools/build/1.0.858#L127

**Comment by usergenic** [comment]
Bunch of things like `Property '_transform' in type 'GulpTransform' is not assignable to the same property in base type 'Transform'.`

**Comment by usergenic** [comment]
Ah this failure is unrelated to branch.  Created a new PR to start to fix the TS errors related to updating package locks at https://github.com/Polymer/tools/pull/309

File: packages/bundler/src/es6-rewriter.ts [hunk 7]
```diff
@@ -432,4 +465,60 @@ export class Es6Rewriter {
...
         {type: 'ImportSpecifier', imported: babel.identifier(exportName)});
   }
+
+  private _rewriteImportMetaToBundleMeta(
+      moduleFile: babel.File,
+      relativeUrl: FileRelativeUrl): babel.File {
+    // Generate a stand-in for any local references to import.meta...
+    // const __bundledImportMeta = {...import.meta, url: __bundledImportURL};
+    // TODO(usergenic): Consider migrating this AST production mishmash into the
+    // `ast` tagged template literal available like this:
```
Comment (justinfagnani) [inline packages/bundler/src/es6-rewriter.ts]: +1. Much easier to read and write.

File: packages/bundler/src/es6-rewriter.ts [hunk 8]
```diff
@@ -432,4 +465,60 @@ export class Es6Rewriter {
...
+    // https://github.com/Polymer/tools/blob/master/packages/build/src/babel-plugin-dynamic-import-amd.ts#L64
+    const bundledImportMetaName = '__bundledImportMeta';
```
Comment (justinfagnani) [inline packages/bundler/src/es6-rewriter.ts]: We should generate this name to be conflict free with identifiers at `import.meta` use sites.

File: packages/bundler/src/es6-rewriter.ts [hunk 9]
```diff
@@ -432,4 +465,60 @@ export class Es6Rewriter {
...
+              babel.identifier(bundledImportMetaName), babel.objectExpression([
+                babel.spreadProperty(babel.memberExpression(
+                    babel.identifier('import'), babel.identifier('meta'))),
+                babel.objectProperty(
+                    babel.identifier('url'),
+                    babel.memberExpression(
+                        babel.newExpression(
+                            babel.identifier('URL'),
+                            [
+                              //
```
Comment (justinfagnani) [inline packages/bundler/src/es6-rewriter.ts]: remove? Or if it's for formatting, move to previous line?

File: packages/bundler/src/es6-rewriter.ts [hunk 10]
```diff
@@ -432,4 +465,60 @@ export class Es6Rewriter {
...
+                              babel.memberExpression(
+                                  babel.memberExpression(
+                                      babel.identifier('import'),
+                                      babel.identifier('meta')),
+                                  babel.identifier('url'))
+                            ]),
+                        babel.identifier('href')))
+              ]))
+        ]);
+    const newModuleFile = clone(moduleFile);
```
Comment (justinfagnani) [inline packages/bundler/src/es6-rewriter.ts]: why do you have to clone it?

File: packages/bundler/src/es6-rewriter.ts [hunk 11]
```diff
@@ -432,4 +465,60 @@ export class Es6Rewriter {
...
+        enter(path: NodePath<babel.MetaProperty>) {
+          const metaProperty = path.node;
+          if (metaProperty.meta.name !== 'import' &&
+              metaProperty.property.name !== 'meta') {
+            // We're specifically looking for instances of `import.meta` so
+            // ignore any other meta properties.
+            return;
+          }
+          const bundledImportMeta = babel.identifier(bundledImportMetaName);
+          rewriteObject(metaProperty, bundledImportMeta);
```
Comment (justinfagnani) [inline packages/bundler/src/es6-rewriter.ts]: use `path.replaceWith` instead?

File: packages/bundler/src/es6-rewriter.ts (hunk 8 above)
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: Added issue. https://github.com/Polymer/tools/issues/310

File: packages/bundler/src/es6-rewriter.ts (hunk 10 above)
Comment (rictic) [reply packages/bundler/src/es6-rewriter.ts]: +1. Modifying documents returned by analyzer will lead to the worst kind of weird bugs with spooky action at a distance.

File: packages/bundler/src/es6-rewriter.ts (hunk 11 above)
Comment (usergenic) [reply packages/bundler/src/es6-rewriter.ts]: Okay. That makes sense here. Replace'd

**Comment by keanulee** [comment]
I'm seeing this transform present on the [es6-bundled](https://github.com/Polymer/pwa-starter-kit/blob/master/polymer.json#L32) build output which doesn't make since it uses `transformModulesToAmd` and makes `import` undefined/an invalid keyword in FF. Is this expected, or is there some other JS config I need in that build configuration @usergenic ?

```js
const bundledImportMeta = { ...import.meta,
  url: new URL('...', import.meta.url).href
};
```

**Comment by usergenic** [comment]
@keanulee the fix was to pin a specific dependency (that was suddenly causing errors after a release a couple days ago) until babel's minify preset is more stable https://github.com/Polymer/tools/pull/380

**Pull request merged**